filters/include/pcl/filters/impl/farthest_point_sampling.hpp

```diff
@@ -0,0 +1,128 @@
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of the copyright holder(s) nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#pragma once
+
+#include <pcl/common/geometry.h>
+#include <pcl/filters/farthest_point_sampling.h>
+#include <pcl/point_cloud.h>
+#include <pcl/point_types.h>
+#include <algorithm>
+#include <limits>
+#include <random>
+
+template<typename PointT> void
+pcl::FarthestPointSampling<PointT>::applyFilter (PointCloud &output)
+{
+  std::vector<int> indices;
+  indices.reserve(sample_);
+  if (keep_organized_)
+  {
+    bool temp = extract_removed_indices_;
+    extract_removed_indices_ = true;
+    applyFilter (indices);
+    extract_removed_indices_ = temp;
+
+    output = *input_;
+    for (const auto ri : *removed_indices_)  // ri = removed index
+      output.points[ri].x = output.points[ri].y = output.points[ri].z = user_filter_value_;
+    if (!std::isfinite (user_filter_value_))
+      output.is_dense = false;
+  }
+  else
+  {
+    output.is_dense = true;
+    applyFilter (indices);
+    pcl::copyPointCloud (*input_, indices, output);
+  }
+}
+
+template<typename PointT> void
+pcl::FarthestPointSampling<PointT>::applyFilter (std::vector<int> &indices)
+{
+  std::size_t size = input_->points.size();
+
+  //check if number of points in cloud is less than requested number of points
+  if (size <= sample_)
+  {
+    PCL_THROW_EXCEPTION (BadArgumentException,
+                        "Requested number of points is greater than point cloud size!");
+    return;
+  }
+
+  int max_index;
+  std::vector<float> distances_to_selected_points (size, std::numeric_limits<float>::max ());
+  
+  //set random seed
+  std::mt19937 random_gen(seed_);
+  std::uniform_int_distribution<> dis(0, size -1);
+
+  //pick the first point at random
+  max_index = dis(random_gen);
+  distances_to_selected_points[max_index] = -1.0;
+  indices.push_back(max_index);
+
+  for (int j = 0; j < sample_ - 1; j++)
```

Comment:
What about [1, sample_) instead of [0, sample_-1)? Doesn't change anything fundamentally, but encourages the reader to notice the iteration counter since the loop doesn't start from 0